src/main/java/com/basho/riak/client/api/GenericRiakCommand.java

```diff
@@ -0,0 +1,87 @@
+package com.basho.riak.client.api;
+
+/**
+ * Created by srg on 11/6/16.
+ */
+
+import com.basho.riak.client.api.commands.CoreFutureAdapter;
+import com.basho.riak.client.core.FutureOperation;
+import com.basho.riak.client.core.PBStreamingFutureOperation;
+import com.basho.riak.client.core.RiakCluster;
+import com.basho.riak.client.core.RiakFuture;
+
+
+public abstract class GenericRiakCommand<R, I, CoreR, CoreI> extends RiakCommand<R, I>
+{
+    @FunctionalInterface
+    protected interface Converter<T, O>
+    {
+        T convert(O source);
+    }
+
+    private Converter<R, CoreR> responseConverter;
+    private Converter<I, CoreI> infoConverter;
+
+    public GenericRiakCommand()
+    {
+        this.responseConverter = this::convertResponse;
+        this.infoConverter = this::convertInfo;
+    }
+
+    public GenericRiakCommand(final Converter<R, CoreR> responseConverter, final Converter<I, CoreI> infoConverter)
+    {
+        this.responseConverter = responseConverter;
+        this.infoConverter = infoConverter;
+    }
+
+    protected abstract FutureOperation<CoreR, ?, CoreI> buildCoreOperation();
+
+    protected RiakFuture<R,I> executeAsync(RiakCluster cluster)
+    {
+        final FutureOperation<CoreR, ?, CoreI> coreOperation = buildCoreOperation();
+        assert coreOperation != null;
+
+        // TODO: WE NEED TO GET RID SUCH A WEIRD IF-FORK
+        final RiakFuture<CoreR, CoreI> coreFuture;
+        if (coreOperation instanceof PBStreamingFutureOperation)
+        {
+            coreFuture = cluster.execute((PBStreamingFutureOperation<CoreR, ?, CoreI>) coreOperation);
+        }
+        else
+        {
+            coreFuture = cluster.execute(coreOperation);
+        }
+
+        assert coreFuture != null;
+
+        final CoreFutureAdapter<R, I, CoreR, CoreI> future =
+                new CoreFutureAdapter<R, I, CoreR, CoreI>(coreFuture)
+                {
+                    @Override
+                    protected R convertResponse(CoreR coreResponse)
+                    {
+                        return responseConverter.convert(coreResponse);
+                    }
+
+                    @Override
+                    protected I convertQueryInfo(CoreI coreQueryInfo)
+                    {
+                        return infoConverter.convert(coreQueryInfo);
+                    }
+                };
+        coreFuture.addListener(future);
+        return future;
+    }
+
+    @SuppressWarnings("unchecked")
```

Comment:
Most of these changes look good (and needed). 
The only thing that stands out are these two `unchecked` suppressions.  We're ignoring the type system here, which always feels bad for some reason.  Could we make these abstract methods, and then provide subclasses where `I == CoreI` and/or `R == CoreR`, and use those where necessary?  

Sort of like what I did with the `ImmediateCoreFutureAdapter`: https://github.com/basho/riak-java-client/pull/677/files#diff-1c6f3b6c65d8e0eeaee247ff938a349bR70 ?